Sources/Model/Quantity.swift

```
@@ -18,21 +18,238 @@ import Foundation
+    DecSuffixed(key: "k", bePair: BePair(base: 10, exponent: 3)),
+    DecSuffixed(key: "M", bePair: BePair(base: 10, exponent: 6)),
+    DecSuffixed(key: "G", bePair: BePair(base: 10, exponent: 9)),
+    DecSuffixed(key: "T", bePair: BePair(base: 10, exponent: 12)),
+    DecSuffixed(key: "P", bePair: BePair(base: 10, exponent: 15)),
+    DecSuffixed(key: "E", bePair: BePair(base: 10, exponent: 18)),
+]
+let decDecimalSI = decDecimalSINag + decDecimalSIPos
+
 
-	public init(integerLiteral value: Int) {
-		self.value = String(value)
-	}
+let frndBinarySI = Array(decBinarySI.reversed())
+var frndDecimalSI = Array((decDecimalSINag.reversed() + decDecimalSIPos).reversed())
 
-	public init(floatLiteral value: Double) {
-		self.value = String(value)
-	}
+
+// TODO: Implement proper quantity serde: https://github.com/kubernetes/apimachinery/blob/master/pkg/api/resource/quantity.go
+public struct Quantity: ExpressibleByStringLiteral, ExpressibleByIntegerLiteral, ExpressibleByFloatLiteral {
+    let value: String
+    
+    var decimalValue: Decimal = 0
+    var unit: String = ""
+    var unitType: UnitType = .decimalSI
+    
+    public func getOriginalValue() -> String {
+        return value
+    }
+    
+    public func getValue() -> Decimal {
+        return decimalValue
+    }
+    
+    public init(stringLiteral value: String) {
+        self.value = value
+        self.parseData(str: value)
+    }
+    
+    public init(integerLiteral value: Int) {
+        self.init(stringLiteral: "\(value)")
+    }
+    
+    public init(floatLiteral value: Double) {
+        self.init(stringLiteral: "\(value)")
+    }
+    
+    mutating func parseData(str: String) {
+        let range = NSRange(location: 0, length: str.count)
+        let regex = try! NSRegularExpression(pattern: "^[0-9.e]+", options: [])
+        let results = regex.matches(in: str, options: [], range: range)
+        if results.count != 1 {
+            return
```

Why have you decided to `return` instead of failing the `init` on invalid input?

---

I have try to throw an error for func `parseData` and `init(stringLiteral:)`, but `ExpressibleByStringLiteral` seems not allow throw error for init(stringLiteral:)

(Or just because my swift is not good enough)

so I add a new property `ok` to validate whether parse data success

Or can I delete these protocols?

---

I would keep the protocols, because they provide a nice syntactic sugar and more importantly would allow us to support a wider range of existing manifests, i.e. many existing manifests define quantities as strings or as numbers etc.

You could throw an error in the `parseData` and in the `init(stringLiteral:)` you can just `fail` with a descriptive message or force-unwrap.